### PR TITLE
Spectral sig defense

### LINKED ISSUE #396: Spectral Signature Backdoor Poisoning Defense
**Is your feature request related to a problem? Please describe.**
Add defense from [Spectral Signatures in Backdoor Attacks](https://papers.nips.cc/paper/8024-spectral-signatures-in-backdoor-attacks.pdf)

**Describe the solution you'd like**
Implementation of the spectral signature defense as a poison filtering defense for neural networks. Should include basic unit tests and necessary documentation.

### PR DESCRIPTION
**Pull request by ng390**
# Description

In this PR:

Fixes #396

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

Several fast-running tests were provided to the test suite. Open to discussion on if a Jupyter notebook would be appropriate
to demonstrate functionality -- it would end up looking quite similar to the poisoning_dataset_mnist.ipynb notebook.

**Test Configuration**:
- OS: Mac OS X
- Python version 3.8.2
- ART version or commit number dev_1.3, branch based off commit hash 53ae3eaa22f25933892baa395d616e7633c54e88
- TensorFlow / Keras

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (related to this feature)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -122,6 +122,7 @@ Get in touch with us on [Slack](https://ibm-art.slack.com) (invite [here](https:
 **Detection of Poisoning Attacks:**
 * Detection based on activations analysis ([Chen et al., 2018](https://arxiv.org/abs/1811.03728))
 * Detection based on data provenance ([Baracaldo et al., 2018](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8473440))
+* Detection based on spectral signatures ([Tran et al., 2018](https://papers.nips.cc/paper/8024-spectral-signatures-in-backdoor-attacks.pdf))
```
Comment (beat-buesser) [inline README.md:125]: Could you please add this addition also to README-cn.md?

File: art/poison_detection/spectral_signature_defense.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        "y_train",
+        "batch_size",
+        "eps_multiplier",
+        "ub_pct_poison",
+        "nb_classes",
+    ]
+
+    def __init__(self, classifier, x_train, y_train, **kwargs):
+        """
+        Create an :class:`.ActivationDefence` object with the provided classifier.
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:51]: Is ActivationDefence still correct?
```suggestion
        Create an :class:`.SpectralSignatureDefense` object with the provided classifier.
```

File: art/poison_detection/spectral_signature_defense.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        :param classifier: Model evaluated for poison.
+        :type classifier: :class:`art.classifiers.Classifier`
+        :param x_train: dataset used to train the classifier.
+        :type x_train: `np.ndarray`
+        :param y_train: labels used to train the classifier.
+        :type y_train: `np.ndarray`
+        """
+        super(SpectralSignatureDefense, self).__init__(classifier, x_train, y_train)
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:59]:
```suggestion
        super().__init__(classifier, x_train, y_train)
```

File: art/poison_detection/spectral_signature_defense.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        :type kwargs: `dict`
+        :return: JSON object with confusion matrix.
+        :rtype: `jsonObject`
+        """
+
+        if is_clean is None or is_clean.size == 0:
+            raise ValueError(
+                "is_clean was not provided while invoking evaluate_defence."
+            )
+        is_clean_by_class = SpectralSignatureDefense.split_by_class(
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:81]:
```suggestion
        is_clean_by_class = self.split_by_class(
```

File: art/poison_detection/spectral_signature_defense.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            is_clean, self.y_train_sparse, self.nb_classes
+        )
+        _, predicted_clean = self.detect_poison()
+        predicted_clean_by_class = SpectralSignatureDefense.split_by_class(
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:85]:
```suggestion
        predicted_clean_by_class = self.split_by_class(
```

File: art/poison_detection/spectral_signature_defense.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+        self.set_params(**kwargs)
+
+        nb_layers = len(self.classifier.layer_names)
+
+        features_x_poisoned = self.classifier.get_activations(
+            self.x_train, layer=nb_layers - 1, batch_size=self.batch_size
+        )
+
+        features_split = SpectralSignatureDefense.split_by_class(
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:114]:
```suggestion
        features_split = self.split_by_class(
```

File: art/poison_detection/spectral_signature_defense.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        "eps_multiplier",
+        "ub_pct_poison",
+        "nb_classes",
+    ]
+
+    def __init__(self, classifier, x_train, y_train, **kwargs):
+        """
+        Create an :class:`.ActivationDefence` object with the provided classifier.
+        :param classifier: Model evaluated for poison.
+        :type classifier: :class:`art.classifiers.Classifier`
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:53]:
```suggestion
        :type classifier: :class:`art.estimators.classification.ClassifierMixin`
```

File: art/poison_detection/spectral_signature_defense.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        self.y_train_sparse = np.argmax(y_train, axis=1)
+
+        self.evaluator = GroundTruthEvaluator()
+
+    def evaluate_defence(self, is_clean, **kwargs):
+        """
+        If ground truth is known, this function returns a confusion matrix in the form of a JSON object.
+        :param is_clean: Ground truth, where is_clean[i]=1 means that x_train[i] is clean and is_clean[i]=0 means
+                         x_train[i] is poisonous.
+        :type is_clean: :class `np.ndarray`
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:70]:
```suggestion
        :type is_clean: `np.ndarray`
```

File: art/poison_detection/spectral_signature_defense.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        return conf_matrix_json
+
+    def detect_poison(self, **kwargs):
+        """
+        Returns poison detected and a report.
+        :return: (report, is_clean_lst):
+                where a report is a dictionary containing the index as keys the outlier score
+                of suspected poisons as values
+                where is_clean is a list, where is_clean_lst[i]=1 means that x_train[i]
+                there is clean and is_clean_lst[i]=0, means that x_train[i] was classified as poison.
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:102]:
```suggestion
                where a report is a dictionary containing the index as keys the outlier score of suspected poisons as
                values where is_clean is a list, where is_clean_lst[i]=1 means that x_train[i] there is clean and
                is_clean_lst[i]=0, means that x_train[i] was classified as poison.
```

File: art/poison_detection/spectral_signature_defense.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            split[int(label)].append(data[idx])
+        return [np.asarray(dat) for dat in split]
+
+    def set_params(self, **kwargs):
+        """
+        Take in a dictionary of parameters and applies defense-specific checks before saving them as attributes.
+        If a parameter is not provided, it takes its default value.
+        """
+        # Save defence-specific parameters
+        super(SpectralSignatureDefense, self).set_params(**kwargs)
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:182]:
```suggestion
        super().set_params(**kwargs)
```

File: art/poison_detection/spectral_signature_defense.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+from art.poison_detection.poison_filtering_defence import PoisonFilteringDefence
+
+
+class SpectralSignatureDefense(PoisonFilteringDefence):
+    """
+    Method from Tran et al., 2018 performing poisoning detection based on Spectral Signatures
+    """
+
+    defence_params = PoisonFilteringDefence.defence_params + [
+        "classifier",
```
Comment (beat-buesser) [inline art/poison_detection/spectral_signature_defense.py:40]: Classifier is alreadu in PoisonFilteringDefence.defence_params
```suggestion
```

File: tests/poison_detection/test_spectral_signature_defense.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+            loss="sparse_categorical_crossentropy",
+            optimizer="adam",
+            metrics=["accuracy"],
+        )
+
+        from art.estimators.classification.keras import KerasClassifier
+
+        cls.classifier = KerasClassifier(model=model, clip_values=(min_, max_))
+
+        cls.classifier.fit(x_train, y_train, nb_epochs=1, batch_size=128)
```
Comment (beat-buesser) [inline tests/poison_detection/test_spectral_signature_defense.py:83]: Would it be possible to use our stnadard Keras classifier for testing with
```suggestion
from tests.utils import get_image_classifier_kr
cls = get_image_classifier_kr()
```

**Comment by ng390** [comment]
Ready for another round of review. Incorporated the suggested changes and resolved the merge conflicts.

**Comment by beat-buesser** [comment]
@ng390 Thank you very much for your pull request contributing the Spectral Signature Backdoor Poisoning Defense!

**Pull request merged**